pkg/storage/azureblob/azureblob.go

```diff
@@ -0,0 +1,163 @@
+	const op errors.Op = "azure.BlobExists"
+	// TODO: Any better way of doing this ?
+	blobURL := c.containerURL.NewBlockBlobURL(path)
+	_, err := blobURL.GetProperties(ctx, azblob.BlobAccessConditions{})
+	if err != nil {
+		var serr azblob.StorageError
+		var ok bool
+
+		if serr, ok = err.(azblob.StorageError); !ok {
+			return false, errors.E(op, fmt.Errorf("Error in casting to azure error type"))
+		}
+		if serr.Response().StatusCode == http.StatusNotFound {
+			return false, errors.E(op, err)
+		}
+
+		return false, errors.E(op, err)
+	}
+	return true, nil
+
+}
+
+// ReadBlob returns an io.ReadCloser for the contents of a blob
+func (c *azureBlobStoreClient) ReadBlob(ctx context.Context, path string) (io.ReadCloser, error) {
+	const op errors.Op = "azure.ReadBlob"
+	blobURL := c.containerURL.NewBlockBlobURL(path)
+	downloadResponse, err := blobURL.Download(ctx, 0, 0, azblob.BlobAccessConditions{}, false)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	return downloadResponse.Body(azblob.RetryReaderOptions{}), nil
+}
+
+func (c *azureBlobStoreClient) ListBlobs(ctx context.Context, prefix string) ([]string, error) {
+	const op errors.Op = "azure.ListBlobs"
+	var blobs []string
+	for marker := (azblob.Marker{}); marker.NotDone(); {
+		listBlob, err := c.containerURL.ListBlobsFlatSegment(ctx, marker, azblob.ListBlobsSegmentOptions{
+			Prefix: prefix,
+		})
+		if err != nil {
+			return nil, errors.E(op, err)
+		}
+		marker = listBlob.NextMarker
+
+		for _, blob := range listBlob.Segment.BlobItems {
+			blobs = append(blobs, blob.Name)
+		}
+	}
+	return blobs, nil
+}
+func (c *azureBlobStoreClient) DeleteBlob(ctx context.Context, path string) error {
+	const op errors.Op = "azure.DeleteBlob"
+	blobURL := c.containerURL.NewBlockBlobURL(path)
+	_, err := blobURL.Delete(ctx, azblob.DeleteSnapshotsOptionNone, azblob.BlobAccessConditions{})
+	if err != nil {
+		return errors.E(op, err)
+	}
+	return nil
+}
+
+// UploadWithContext uploads a blob to the container
+func (c *azureBlobStoreClient) UploadWithContext(ctx context.Context, path, contentType string, content io.Reader) error {
+	const op errors.Op = "azure.UploadWithContext"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	blobURL := c.containerURL.NewBlockBlobURL(path)
+	emptyMeta := map[string]string{}
+	emptyBlobAccessCond := azblob.BlobAccessConditions{}
+	httpHeaders := func(contentType string) azblob.BlobHTTPHeaders {
+		return azblob.BlobHTTPHeaders{
+			ContentType: contentType,
+		}
+	}
+	bufferSize := 1 * 1024 * 1024 // Size of the rotating buffers that are used when uploading
+	maxBuffers := 3               // Number of rotating buffers that are used when uploading
+
+	uploadStreamOpts := azblob.UploadStreamToBlockBlobOptions{
+		BufferSize:       bufferSize,
+		MaxBuffers:       maxBuffers,
+		BlobHTTPHeaders:  httpHeaders(contentType),
+		Metadata:         emptyMeta,
```

Comment:
If we don't set any value in them, this properties ( `Metadata`, `AccessConditions` ) can be omitted because they will be set to their coresponding `Nil Values`.